apollo-ios/Design/swift-identifiable.md

```diff
@@ -0,0 +1,94 @@
+
+Apollo Client iOS could assist the developer by adding conformance to the Identifiable protocol to its generated models. Selecting the field to be used as an identity is done by adding a new GraphQL directive.
+
+# Definition
+
+The directive is defined as:
+```graphql
+directive @apollo_client_ios_identity on FIELD | FIELD_DEFINITION
+```
+
+This directive MUST be used on a field with a non-nullable scalar type. 
+
+## Usage in types
+
+A type MAY have a field annotated with the directive.
+
+```graphql
+type Animal {
+	id: ID! @apollo_client_ios_identity
+	name: String
+}
+```
+
+Within a type, the directive MUST NOT be used on more than one field.
+
+## Usage in interfaces
+
+It's possible for interfaces to use the directive on a field.
+
+```graphql
+interface Identifiable {
+	id: ID! @apollo_client_ios_identity
+}
+```
+
+Implementations of this interface MUST copy the directive to the same field.
+
+## Usage in operations
+
+It's likely that an external schema will not use this directive. In this case, an identity MAY be chosen when writing a query. The directive does not need to be included in the query sent to the GraphQL server.
+
+```graphql
+query GetAllAnimals {
+	allAnimals {
+		id @apollo_client_ios_identity
+		string
+	}
+}
+```
+
+It is allowed for a query and a type to both use the directive to describe the same field.
+
+If the server defines a field as an identity, a query SHOULD NOT choose another field. A query MUST NOT use the directive on more than one field in the same selection (unless they are in differently nested objects).
+
+## Generated code
+
+If an operation contains a field marked with the `@apollo_client_ios_identity` directive (by either the schema or the operation itself), the generated SelectionSet will have a conformance to the Identifiable protocol.
+
+```swift
+// Inline selection
+public struct Data: AnimalKingdomAPI.SelectionSet, Identifiable { /* ... */ }
+
+// Fragment selection
+public struct PetDetails: AnimalKingdomAPI.SelectionSet, Fragment, Identifiable { /* ... */ }
+```
+
+The protocol requires that the identity is accessible through a public field named exactly `id`. If the annotated field has a different name, an additional getter will be generated:
+
+```swift
+public var id: String { self.uuid }
+```
+
+## Naming conflicts
+
+If the identity field is not called `id`, but another field called `id` is present in the selection, a custom getter cannot be added. Swift does not support using another field to handle the conformance.
+
+In this case, a conformance to Identifiable SHOULD NOT be generated.
```

Comment:
In this case we should definitely be throwing an error to notify the user of the issue, whether or not the error should actually cause code generation to fail would be the other question. I think it's acceptable for the error to be thrown and logged without failing code generation but could be persuaded either way.

Reply:
+1 to validation.

Not adding conformance to `Identifiable` will not prevent the generated code from building. It will certainly be confusing to the user but not result in a build error so I prefer logging over failing.